File: DS_basics/tree/huffman/huffmancode/huffmanEnDecode.c
#include <stdio.h>
#include <stdlib.h>
#include <string.h>

typedef struct HuffmanNode
{
    char value;
    int freq;
    int index;
    struct HuffmanNode *parent, *left, *right;
} HuffmanNode;

HuffmanNode *nodes_list[512];
int nodes_count = 0;

HuffmanNode *createNode(char value, int freq)
{
    HuffmanNode *node = (HuffmanNode *)malloc(sizeof(HuffmanNode));
    node->value = value;
    node->freq = freq;
    node->index = nodes_count;
    nodes_list[nodes_count++] = node;
    node->parent = node->left = node->right = NULL;
    return node;
}

int comparator(const void *p, const void *q)
{
    return ((HuffmanNode *)p)->freq - ((HuffmanNode *)q)->freq;
}
HuffmanNode *buildHuffmanTree(const char *data)
{
    int freq[256] = {0};
    int i;
    for (i = 0; data[i]; i++)
    {
        freq[(unsigned char)data[i]]++;
    }

    HuffmanNode *heap[256];
    int heapSize = 0;
    for (i = 0; i < 256; i++)
    {
        if (freq[i] > 0)
        {
            heap[heapSize++] = createNode(i, freq[i]);
        }
    }

    while (heapSize > 1)
    {
        qsort(heap, heapSize, sizeof(HuffmanNode *), comparator); // 这一步是关键，每次都要把频率最小的两个节点放在最前面

        HuffmanNode *left = heap[0];
        HuffmanNode *right = heap[1];
        for (i = 0; i < heapSize - 2; i++)
        {
            heap[i] = heap[i + 2];
        }
        heapSize -= 2;

        HuffmanNode *parent = createNode(0, left->freq + right->freq);
        parent->left = left;
        parent->right = right;
        left->parent = parent;
        right->parent = parent;

        printf("合并节点 %d(%c:%d) 和 %d(%c:%d)，生成新节点 %d(%d)\n",
               left->index, left->value ? left->value : ' ', left->freq,
               right->index, right->value ? right->value : ' ', right->freq,
               parent->index, parent->freq);

        heap[heapSize++] = parent;
    }

    return heap[0];
}

void printHuffmanTree()
{
    printf("---------------------------------------------------------------------\n");
    printf("| Index |  Node  | Weight | Parent | Left Child | Right Child |\n");
    printf("---------------------------------------------------------------------\n");
    int i;

    for (i = 0; i < nodes_count; i++)
    {
        char nodeStr = nodes_list[i]->value ? nodes_list[i]->value : ' ';
        int parentIndex = nodes_list[i]->parent ? nodes_list[i]->parent->index : -1;
        int leftIndex = nodes_list[i]->left ? nodes_list[i]->left->index : -1;
        int rightIndex = nodes_list[i]->right ? nodes_list[i]->right->index : -1;
        printf("|%7d|%8c|%8d|%8d|%12d|%13d|\n",
               nodes_list[i]->index, nodeStr, nodes_list[i]->freq, parentIndex, leftIndex, rightIndex);
    }
    printf("---------------------------------------------------------------------\n");
}

void buildHuffmanTable(HuffmanNode *node, char *code, char table[256][256])
{
    if (node->value)
    {
        strcpy(table[(unsigned char)node->value], code);
        return;
    }
    if (node->left)
    {
        char newCode[256];
        sprintf(newCode, "%s0", code);
        buildHuffmanTable(node->left, newCode, table);
    }
    if (node->right)
    {
        char newCode[256];
        sprintf(newCode, "%s1", code);
        buildHuffmanTable(node->right, newCode, table);
    }
}

char *encode(const char *data, char table[256][256])
{
    static char encoded[4096];
    encoded[0] = '\0';
    int i;
    for (i = 0; data[i]; i++)
    {
        strcat(encoded, table[(unsigned char)data[i]]);
    }
    return encoded;
}

char decodeChar(HuffmanNode *root, const char **pencoded)
{
    HuffmanNode *node = root;
    while (!node->value && **pencoded)
    {
        node = **pencoded == '0' ? node->left : node->right;
        (*pencoded)++;
    }
    return node->value;
}

char *decode(const char *encoded, HuffmanNode *root)
{
    static char decoded[1024];
    char *pdecoded = decoded;
    while (*encoded)
    {
        *pdecoded++ = decodeChar(root, &encoded);
    }
    *pdecoded = '\0';
    return decoded;
}
int main()
{
    const char data[] = "This is a test message for Huffman coding.";
    HuffmanNode *root = buildHuffmanTree(data);

    char table[256][256] = {0};
    buildHuffmanTable(root, "", table);

    char *encodedData = encode(data, table);
    char *decodedData = decode(encodedData, root);

    printHuffmanTree();
    printf("原始数据：%s\n", data);
    printf("编码后的数据：%s\n", encodedData);
    printf("解码后的数据：%s\n", decodedData);

    return 0;
}

File: DS_basics_no_visualizations/tree/huffman/huffmancode/huffmanEnDecode.py
from heapq import heappush, heappop
from collections import defaultdict


class HuffmanNode:
    def __init__(self, value=None, freq=0, left=None, right=None):
        self.value = value
        self.freq = freq
        self.left = left
        self.right = right
        self.index = -1  # 初始化index为-1

    def __lt__(self, other):
        return self.freq < other.freq


def build_huffman_tree(data):
    # 统计字符频率
    freqs = defaultdict(int)
    for char in data:
        freqs[char] += 1

    # 初始化节点列表
    nodes_list = []

    # 构建哈夫曼树
    heap = [HuffmanNode(char, freq) for char, freq in freqs.items()]
    for i in range(len(heap)):
        heap[i].index = i
        nodes_list.append(heap[i])  # 添加到节点列表
    while len(heap) > 1:
        left = heappop(heap)
        right = heappop(heap)
        parent = HuffmanNode(freq=left.freq + right.freq,
                             left=left, right=right)
        parent.index = len(nodes_list)  # 使用节点列表的长度作为索引
        heappush(heap, parent)
        nodes_list.append(parent)  # 添加新生成的节点到列表

        # 打印信息
        print(f"合并节点 {left.index}({left.value}:{left.freq}) 和 {right.index}({right.value}:{right.freq})，生成新节点 {parent.index}({parent.freq})")

    # 打印哈夫曼树表格
    print_huffman_tree(heap[0])

    # 返回哈夫曼树的根节点
    return heap[0]


def build_huffman_table(root):
    # 构建哈夫曼编码表
    table = {}
    stack = [(root, "")]
    while stack:
        node, code = stack.pop()
        if node.value is not None:
            table[node.value] = code
        if node.left is not None:
            stack.append((node.left, code + "0"))
        if node.right is not None:
            stack.append((node.right, code + "1"))
    return table


def encode(data, table):
    # 哈夫曼编码
    encoded = "".join(table[char] for char in data)

    # 打印信息
    print(f"原始数据：{data}")
    print(f"编码后的数据：{encoded}")

    return encoded


def decode(data, root):
    # 哈夫曼解码
    result = []
    node = root
    for bit in data:
        if bit == "0":
            node = node.left
        else:
            node = node.right
        if node.value is not None:
            result.append(node.value)
            node = root

    # 打印信息
    decoded = "".join(result)
    print(f"解码后的数据：{decoded}")

    return decoded

def print_huffman_tree(root):
    # 打印哈夫曼树表格

    # 收集所有节点
    nodes = []
    stack = [(root, root.freq, "", "")]
    while stack:
        node, weight, parent, direction = stack.pop()
        if node is not None:
            nodes.append(node)
            stack.append((node.left, node.left.freq if node.left else 0, str(node.index), direction + "0"))
            stack.append((node.right, node.right.freq if node.right else 0, str(node.index), direction + "1"))

    # 按索引排序
    nodes.sort(key=lambda x: x.index)

    print("-" * 69)
    print("| Index |  Node  | Weight | Parent | Left Child | Right Child |")
    print("-" * 69)

    for node in nodes:
        if node.value is not None:
            node_str = f"'{node.value}'"
        else:
            node_str = ""
        left_str = str(node.left.index) if node.left else ""
        right_str = str(node.right.index) if node.right else ""
        print(f"|{node.index:^7}|{node_str:^8}|{node.freq:^8}|{'':^8}|{left_str:^12}|{right_str:^13}|")

    print("-" * 69)



# 测试代码
data = "This is a test message for Huffman coding."
root = build_huffman_tree(data)
table = build_huffman_table(root)
encoded = encode(data, table)
decoded = decode(encoded, root)
